teps/0120-canceling-concurrent-pipelineruns.md

```diff
@@ -102,7 +102,326 @@ queueing, or more advanced concurrency strategies.
+  strategy: cancelRunFinally
+---
+kind: Trigger
+  name: github-pr-trigger
+spec:
+  template:
+    ref: github-template
+  bindings:
+  - name: pull-request-id
+    value: $(body.pull-request.id)
+  concurrency:
+    ref: pull-requests
+```
+
+A ConcurrencyControl would apply to any PipelineRuns in that namespace, and its parameters would be substituted with the PipelineRun's parameters.
+
+Here, the Triggers controller (or a mutating admission webhook) would create all PipelineRuns as "pending", and label them with the ConcurrencyControl
+they're referencing (for example: `tekton.dev/concurrencyControl: pull-requests`).
+The PipelineRun reconciler (or a new ConcurrencyControl reconciler) would be responsible for patching PipelineRuns to start or cancel them,
+and reconciling ConcurrencyControls. It would find the concurrency control specified in the PipelineRun label, calculate the concurrency key, and apply
+the key as a label to the PipelineRun (for example: `tekton.dev/concurrencyKey: 1234`). It would then query for any PipelineRuns with the same key
+and cancel all but the most recently started.
+
+There are two approaches we could choose to take when a ConcurrencyControl is added or updated. The first approach, which is simpler, is to decide
+that new/updated ConcurrencyControls do not apply to currently running PipelineRuns. If we choose this approach, a
+ConcurrencyControl CRD doesn’t add much compared to a ConfigMap.
+
+The other possible approach is to update PipelineRun concurrency keys whenever there’s an event related to a ConcurrencyControl.
+In this approach, we’d add a custom handler to enqueue all running PipelineRuns when a ConcurrencyControl is updated, since the ConcurrencyControl doesn't "know"
+what PipelineRuns it's responsible for. This approach means we’d have to recalculate a PipelineRun’s concurrency keys and cancel matching PipelineRuns
+on each reconcile loop. In this scenario, it’s also not guaranteed that PipelineRuns get requeued in any particular order, so we would need to make sure that any PipelineRuns being canceled started before the one being reconciled. 
+
+Lastly, we wouldn't be able to prevent users from manipulating PipelineRun concurrency by editing the `tekton.dev/concurrencyControl` label.
+
+The appeal of a solution involving a separate CRD is that any higher-level controller (e.g. Triggers, Workflows, Pipelines as Code) could get concurrency controls
+for "free" by creating a ConcurrencyControl, and handing the logic off to a separate controller.
+However, in practice, it's difficult to do this in a way that achieves good separation of concerns between reconcilers. Because the PipelineRun controller
+(or ConcurrencyControl controller) is not responsible for creating PipelineRuns, it has to rely on other components starting PipelineRuns as pending, and matching
+these PipelineRuns to the ConcurrencyControl via some strategy such as labels, ownerReferences, or modifying PipelineRuns themselves.
```

Comment:
> Because the PipelineRun controller
(or ConcurrencyControl controller) is not responsible for creating PipelineRuns, it has to rely on other components starting PipelineRuns as pending, and matching
these PipelineRuns to the ConcurrencyControl via some strategy such as labels, ownerReferences, or modifying PipelineRuns themselves

Althought I slightly agree, the `PipelineRun` reconciler has **all** the possibilities to do whatever it wants with the newly created `PipelineRun`. We could decide that, if a "conccurency" label is set on a PipelineRun, we make it pending. This is a choice we can make (or not).